activerecord/CHANGELOG.md

```diff
@@ -1,3 +1,21 @@
+*   Allow MySQL warnings to be reported.
+
+    Active Record configs can be set to enable MySQL warning reporting.
+
+    ```ruby
+    # Configure action to take when SQL query produces warning
+    config.active_record.db_warnings_action = :raise
+
+    # Configure allowlist of warnings that should always be ignored
+    config.active_record.db_warnings_ignore = [
+      /Invalid utf8mb4 character string/,
+      "An exact warning message",
+      ->(warning, db_config) { warning.code == 1292 || db_config.name == "foo" },
```

Comment:
I think I'd rather have a method on the db_config, like `db_config.emit_db_warnings` that's defined in the database.yml per config rather than checking the name specifically.

Reply:
Ooh interesting 🤔 I guess the tradeoff with making it a db config is that it's either / or -- for a particular connection, you're either emitting warnings or not. If we pass the `db_config` here, it gives users more flexibility. Using `db_config.name == "foo"` was a super simple example, but suppose someone wanted to make the condition more specific:
```ruby
->(warning, db_config) do
  # Only ignore "Some message" warnings on connection with db config name "foo" 
  db_config.name == "foo" && warning.message.match?(/Some message/)
end
```
But maybe that's an unnecessary level of granularity at this point. What do you think?

Reply:
I'm realizing now that we could just make the argument to move the logic to `db_warnings_action` if this level of detail is needed -- e.g.
```ruby
db_warnings_action = ->(warning, db_config) do
  return if db_config.name == "foo" && warning.message.match?(/Some message/)

  # Raise or whatever
end
```

So the API would be:
- Action to take on warning is configurable per connection by passing `db_config` into action proc
- `db_warning_ignore` can specify a proc that receives only the `SqlWarning` object itself
- In the `database.yml`, warnings for individual connections can be ignored by setting `emit_db_warnings: false`

Reply:
Sounds good. TBH I'm not even sure the database name or anything other than the warning is really needed. 

I suggested to allow per database configuration, but is there really an use case where you have two databases but want to filter warnings differently on these?

Reply:
My only thought would be if 1) you have different adapters involved, and 2) their warnings are slightly different object shapes, and 3) you want to write your filtering logic without having to account for that differing method support.

Agree that doesn't seem like a particularly compelling use case. 

My original suggestion was definitely focused on perhaps wanting to use different symbol values for different DBs (e.g. raising on warnings inside the "owned" database, but ignoring warnings from some secondary DB that belongs to another app)... the proc case is much trickier, and I don't see the need for varying that level of control.

That being said, if we want to pass any additional value, maybe we should just pass along the actual connection instance? That's already a user-facing object, and is all the detail we have, so no chance we'll later wish we'd exposed a bit more / have to add another param.

Reply:
> maybe we should just pass along the actual connection instance? That's already a user-facing object

Makes sense to me.